### PR TITLE
Add constants for the DCA sorting modes and flags

### PR DESCRIPTION
**Pull request by bezin**
This PR adds constants for sorting mode and flags, that can be used in the data container arrays. This gives us a little bit more autocomplete in IDEs and does not need us to remember every single mode and flag or look it up in the docs every time ;-)

### CODE REVIEW

**Comment by m-vo** [comment]
That's a really helpful addition!

I'd put it into `DC_Table` though, as it's only relevant there.  E.g. with a prefix similar to the config key name: `SORTING_MODE__TREE`, `SORTING_FLAG__INITIAL_LETTER_ASC` and so on.

**Comment by fritzmg** [comment]
I prefer having separate classes, rather than packing everything into `DC_Table`.

**Comment by m-vo** [comment]
> I prefer having separate classes, rather than packing everything into DC_Table.

But then you need to know the classes…

**Comment by bezin** [comment]
For me typing `SortingMode` or `SortingFlag` came more naturally when I am in the  DCA `sorting` section configuring `mode` or `sorting`. Also, I did not have a Kneifzange at hand to work on `DC_Table` 😛 

I would prefer this: 


over this:

**Comment by fritzmg** [comment]
> But then you need to know the classes…

But that's also the case with `DC_Table`. You don't write DCAs within the `DC_Table` or `DataContainer` class, so you wouldn't be able to just type `self::` anyway. You would need to have knowledge about where the flags are stored in any case.

**Comment by aschempp** [comment]
This way, I have to remember 1,2,3 different class names! I would prefer it in `DataContainer` as well, e.g. `DataContainer::SORT_ INITIAL_LETTER_ASC`.

**Comment by rabauss** [comment]
I'm torn! On the one hand, you have to remember the classes, but on the other hand, what are the other large lists of constants for DCAs? Actually only the input types, right? If we add it to the `DataContainer` it should be added via traits!

**Comment by bezin** [comment]
Yes, good point. I would like to add these as well. Wouldn't it make sense, to add the input type constants to every widget class? Which, I must admit, would make the case for adding the sorting mode and flag constants to the DataContainer/DC_Table 🤔

**Comment by leofeyer** [comment]
> This way, I have to remember 1,2,3 different class names! I would prefer it in `DataContainer`

Me too.

**Comment by aschempp** [comment]
> Wouldn't it make sense, to add the input type constants to every widget class? 

It would actually make sense to use the class name instead of an arbitrary string for the widget types… not sure if we can add that in a BC-way 🤔

**Comment by fritzmg** [comment]
> It would actually make sense to use the class name instead of an arbitrary string for the widget types…

I thought we did not want to support this: https://github.com/contao/contao/pull/2378#issuecomment-714511735

**Comment by bezin** [comment]
I deleted the new classes and moved everything to `DC_Table` – as @m-vo already mentioned, the constants are only used there but not in `DC_Folder` or `DC_File`.

**Comment by leofeyer** [comment]
> I'd put it into `DC_Table` though, as it's only relevant there. E.g. with a prefix similar to the config key name: `SORTING_MODE__TREE`, `SORTING_FLAG__INITIAL_LETTER_ASC` and so on.

I don't like this. Yes, the constants might not be used in DC_Folder, but they might be used in a custom child class of DC_Folder. In the end, it is more about in the scope in which the constants make sense – and that is definitely in any data container. Thus `DataContainer::SORT_ASC` instead of `DC_Table::SORT_ASC`. (Files could be sorted descending, too.)

**Comment by leofeyer** [comment]
As discussed in Mumble on April 8th, two things should be changed:

1. The constants should be in the `DataContainer` class.
2. The constants should be used in the code. 😄

**Comment by bezin** [comment]
Hello there, thanks for your comments (unfortunately I could not make it to the Mumble call as I was on my holidays). 
I updated as you discussed. I hope I got them all – please, somebody double check :-)

One minor thing came up: The sorting flags `3` and `4` differ a little between the list sorting flag and the per-field sorting flag: For the list it is always `Sort by initial first two letters` (hence the name), but for a single field you may configure how many letters should be taken into account with the `length` option.

I still used the sorting flag there, as I thought everything else would be more confusing. If somebody has an idea for a name that reflects both possibilities, we could update the naming.

**Comment by bezin** [comment]
Hi there, I actually thought that this was finished already. 

Anyways, I renamed `SORT_INITIAL_TWO_LETTERS_ASC` to `SORT_INITIAL_LETTERS_ASC` so we now have the following options:

* **SORT_INITIAL_LETTER_ASC (or_DESC)**
  Sort by the initial letter
* **SORT_INITIAL_LETTERS_ASC (or _DESC)**
   Sort by initial letters, which is either fixed to the first two characters in list sorting or can be configured with `length` when used in a field configuration.

It's final from my side now, let me know if there is anything I can do.

Cheers 👋

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 1]
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
...
 
+	/**
+	 * Records are not sorted
+	 */
+	public const MODE_UNSORTED = 0;
+
+	/**
+	 * Records are sorted by a fixed field
+	 */
+	public const MODE_FIXED_FIELD = 1;
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
	public const MODE_SORTED = 1;
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 2]
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
...
+
+	/**
+	 * Records are sorted by a switchable field
+	 */
+	public const MODE_SWITCHABLE_FIELD = 2;
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
	public const MODE_SORTABLE = 2;
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 3]
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
...
+
+	/**
+	 * Records are sorted by the parent table
+	 */
+	public const MODE_PARENT_TABLE = 3;
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
	public const MODE_SORT_PARENT = 3;
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 4]
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
...
+
+	/**
+	 * Displays the child records of a parent record (see style sheets module)
+	 */
+	public const MODE_CHILD_RECORD = 4;
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
	public const MODE_CHILDREN = 4;
```

File: core-bundle/src/Resources/contao/classes/DataContainer.php [hunk 5]
```diff
@@ -143,6 +143,101 @@ abstract class DataContainer extends Backend
...
+	 */
+	public const MODE_SWITCHABLE_FIELD = 2;
+
+	/**
+	 * Records are sorted by the parent table
+	 */
+	public const MODE_PARENT_TABLE = 3;
+
+	/**
+	 * Displays the child records of a parent record (see style sheets module)
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/classes/DataContainer.php]:
```suggestion
	 * Displays the child records of a parent record (see content elements)
```

**Comment by bezin** [comment]
I do like all of your suggestions 👍 Will update them later.

**Comment by bezin** [comment]
Everything is updated.

**Review by aschempp — DISMISSED** [review]

**Review by leofeyer — CHANGES_REQUESTED** [review]
I am not happy with `MODE_CHILDREN` and `MODE_CHILD_TREE`. Can we come up with something else? @contao/developers /cc

**Comment by leofeyer** [comment]
@bezin Can you please rebase the PR on the current `4.x` branch and resolve the conflicts?

**Comment by bezin** [comment]
What about `MODE_CHILD_TREE` and `MODE_CHILD_LIST` instead of `MODE_CHILDREN`?

**Review by aschempp — CHANGES_REQUESTED** [review]
```php
public const MODE_UNSORTED = 0;
public const MODE_SORTED = 1;
public const MODE_SORTABLE = 2;
public const MODE_SORTED_PARENT = 3;
public const MODE_PARENT = 4;
public const MODE_TREE = 5;
public const MODE_TREE_EXTENDED = 6;
```

**Comment by bezin** [comment]
All done here from my point of view 👍

**Review by Toflar — DISMISSED** [review]
Nice work @bezin <3

**Comment by Toflar** [comment]
Need to fix the CI tests though ;)

File: core-bundle/src/Resources/contao/drivers/DC_Table.php [hunk 6]
```diff
@@ -4800,7 +4800,7 @@ protected function listView()
...
-<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < 4) ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a> ' : '') . $this->generateGlobalButtons() . '
+<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < self::MODE_PARENT ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a> ' : '') . $this->generateGlobalButtons() . '
```
Comment (bytehead) [inline core-bundle/src/Resources/contao/drivers/DC_Table.php]: There's a `)` missing after `self::MODE_PARENT`

**Review by bytehead — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/drivers/DC_Table.php (hunk 6 above)
Comment (bytehead) [reply core-bundle/src/Resources/contao/drivers/DC_Table.php]:
```suggestion
<a href="' . ($this->ptable ? $this->addToUrl('act=create' . ((($GLOBALS['TL_DCA'][$this->strTable]['list']['sorting']['mode'] ?? null) < self::MODE_PARENT) ? '&amp;mode=2' : '') . '&amp;pid=' . $this->intId) : $this->addToUrl('act=create')) . '" class="header_new" title="' . StringUtil::specialchars($labelNew[1] ?? '') . '" accesskey="n" onclick="Backend.getScrollOffset()">' . $labelNew[0] . '</a> ' : '') . $this->generateGlobalButtons() . '
```

Comment (bezin) [reply core-bundle/src/Resources/contao/drivers/DC_Table.php]: Thanks, just spotted this myself 👍

**Comment by bezin** [comment]
Done 👍

**Review by bytehead — DISMISSED** [review]

**Comment by leofeyer** [comment]
I would also change this: https://github.com/bezin/contao/pull/1

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bezin.

**Pull request merged**